### ISSUE TITLE
TODO list (updated 2026/02/17)

### ISSUE THREAD

**Issue by learning-chip** [huawei-csl/pto-dsl#1]
### Critical missing feature

- [x] Use `scf.ForOp` and `scf.IfOp` to express **dynamic shape** kernel, ref [tmatmulk.py](https://github.com/zhangstevenunity/PTOAS/blob/b3547faad48384705e422bee710416d89b4fe290/test/samples/MatMul/tmatmulk.py#L176)  
  - Requirement: it should take arbitrary-shape tensor just like [this dynamic silu example](https://github.com/huawei-csl/pto-kernels/pull/6)
  - update: done by #3 #4 
  - [x] Deal with masking & valid-shape in later separate PR, done in -
    - https://github.com/huawei-csl/pto-dsl/pull/11
    - https://github.com/huawei-csl/pto-dsl/pull/12
  - [x] Syntax sugar: [`scf.for_`](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.7/mlir/python/mlir/dialects/scf.py#L106) is a built-in syntax sugar for `scf.ForOp` region, done in https://github.com/huawei-csl/pto-dsl/pull/9

### Generalization of DSL wrapper APIs

- [ ] Add Pythonic wrapper and unit tests to cover **all operations** used in [PTOAS/test/samples](https://github.com/zhangstevenunity/PTOAS/tree/main/test/samples) and [pto-isa/tests/npu](https://gitcode.com/cann/pto-isa/tree/master/tests/npu/a2a3/src/st/testcase)
  - Requirement: must **verify numerical accuracy on-device**.
  - All [torch activation functions](https://docs.pytorch.org/docs/stable/nn.functional.html#non-linear-activation-functions) are good examples for composing vector ops. Can test multiple ISA invocation inside one kernel, to save time and reduce code length. 
  - first Vec, then Cube, then Mix kernels
    - [ ] test all vector operations
      - https://github.com/huawei-csl/pto-dsl/pull/10
      - https://github.com/huawei-csl/pto-dsl/pull/20
    - [ ] test all Cube operations
      - https://github.com/huawei-csl/pto-dsl/pull/13
      - https://github.com/huawei-csl/pto-dsl/pull/19
    - [ ] test a few mix kernels such as FA, dequant-matmul
      - a good test list is https://github.com/zhangstevenunity/PTOAS/pull/63 , the frontend should generate all the IRs in there
    - (this task is very suitable for AI agents...)

(@MirkoDeVita98 ?)

- [ ] API doc for the **complete list of supported operations**, using standard doc tools like [mkdocstrings-python](https://mkdocstrings.github.io/python/)
- [ ] User-training examples like [tilelang-puzzles](https://github.com/tile-ai/tilelang-puzzles)

### Applications

- [ ] Port aiv-inv, aic-inv, fast-hadamard from cpp to Python, use auto-sync from ptoas
- [ ] **Performance measurement compared to hardware peak**, documented in `./doc` directory
- [ ] More algorithms such as e2e GDN with chunk 128

(@zouzias, @asobczyk, @gioelegott )

- [ ] **Automatic, massive porting** from Triton kernels like [flash-linear-attention](https://github.com/fla-org/flash-linear-attention), [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) and CuTile kernels like [TileGym](https://github.com/NVIDIA/TileGym), providing skills/knowledge database like [in HF kernels](https://github.com/huggingface/kernels/tree/5ac28bf4a39ca239e698fabee6d19db6b358cf52/skills/cuda-kernels) and [in flash-infer](https://github.com/flashinfer-ai/flashinfer/tree/952b6ab2838d676b4257fcc23bb00f67fdd38efc/.claude/skills)

### JIT compiler enhancement

- [ ] Frontend style improvement: might reuse some of [mlir-python-extras](https://github.com/makslevental/mlir-python-extras) & [eudsl](https://github.com/llvm/eudsl/tree/main/projects/eudsl-python-extras) 
   - (`func`, `scf`, `arith` are most relevant. But `tensor`, `memref`, `linalg` are not used at all. Better cherry-pick useful parts and cite original source)
   - Use ast-transform only very prudently, only until you have to...
- collect more ideas to https://github.com/huawei-csl/pto-dsl/issues/6

(@filipzrc ?)

- [ ] clear option to 1) jit-compile, 2)reuse-cached, or 3) aot-compile kernels
- [x] optionally allow manual-mode without auto-sync (only when auto-sync can't get peak perf)
- [ ] Also test on A5 backend and adapt frontend accordingly (need real hardware...

### CI/CD

- [x] Add a top-level `examples/all_test.sh` or `pytest` entry for the example directory, to auto-manually check each PR
- [x] GitHub actions CI for all non-device unit tests
- [ ] Pip wheel release (not urgent)
  - better bundle with ptoas wheel so that `pip install pto-dsl` brings both
